File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0)
            return 0;

        HashSet<Character> hm = new HashSet<>();
        int maxLength = 0;
        int l = 0;

        for (int r = 0; r < s.length(); r++) {
            while (hm.contains(s.charAt(r))) {
                hm.remove(s.charAt(l));
                l++;
            }
            hm.add(s.charAt(r));
            maxLength = Math.max(maxLength, r - l + 1);
        }

        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dup = set()
        lft=0
        rgt=1
        
        max_lenght=0
        s_length=len(s)
        if s_length==1:
            return 1
        if s_length==0 :
            return 0
        dup.add(s[0])
        while(rgt<s_length):
            if s[rgt] not in  dup:
                dup.add(s[rgt])
                word_len=rgt-lft+1
                rgt=rgt+1
                max_lenght = max(max_lenght, word_len)
            else:
                dup.discard(s[lft])
                lft=lft+1
           
        return max_lenght


            

           



        